File: system/ulib/fs/service.cpp
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <fs/service.h>

namespace fs {

Service::Service(Connector connector)
    : connector_(fbl::move(connector)) {}

Service::~Service() = default;

zx_status_t Service::ValidateFlags(uint32_t flags) {
    if (flags & ZX_FS_FLAG_DIRECTORY) {
        return ZX_ERR_NOT_DIR;
    }
    return ZX_OK;
}

zx_status_t Service::Getattr(vnattr_t* attr) {
    // TODO(ZX-1152): V_TYPE_FILE isn't right, we should use a type for services
    memset(attr, 0, sizeof(vnattr_t));
    attr->mode = V_TYPE_FILE;
    attr->nlink = 1;
    return ZX_OK;
}

zx_status_t Service::Serve(Vfs* vfs, zx::channel channel, uint32_t flags) {
    ZX_DEBUG_ASSERT(!(flags & ZX_FS_FLAG_DIRECTORY)); // checked by Open

    if (!connector_) {
        return ZX_ERR_NOT_SUPPORTED;
    }
    return connector_(fbl::move(channel));
}

} // namespace fs

File: sdk/dart/fuchsia_vfs/lib/src/service.dart
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:fidl/fidl.dart' as fidl;
import 'package:fidl_fuchsia_io/fidl_async.dart';
import 'package:zircon/zircon.dart';

import 'vnode.dart';

// ignore_for_file: unnecessary_null_comparison

/// Function passed by service implementation which can be
/// used to serve connection over [request].
typedef Connector<T> = void Function(fidl.InterfaceRequest<T> request);

/// A node which binds a channel to a service implementation when opened.
class Service<T> extends Vnode {
  final Connector<T> _connector;
  bool _closed = false;

  /// Constructor with [Connector]
  Service.withConnector(this._connector) : assert(_connector != null);

  @override
  int inodeNumber() {
    return inoUnknown;
  }

  @override
  int connect(OpenFlags flags, int mode, fidl.InterfaceRequest<Node> request,
      [OpenFlags? parentFlags]) {
    if (_closed) {
      sendErrorEvent(flags, ZX.ERR_NOT_SUPPORTED, request);
      return ZX.ERR_NOT_SUPPORTED;
    }
    final status = _validateFlagsAndMode(flags, mode);
    if (status != ZX.OK) {
      sendErrorEvent(flags, status, request);
      return status;
    }
    _connector(fidl.InterfaceRequest(request.passChannel()));

    return ZX.OK;
  }

  @override
  DirentType type() {
    return DirentType.service;
  }

  int _validateFlagsAndMode(OpenFlags flags, int mode) {
    if ((mode & ~modeProtectionMask) & ~modeTypeService != 0) {
      return ZX.ERR_INVALID_ARGS;
    }
    if (flags & OpenFlags.directory != OpenFlags.$none) {
      return ZX.ERR_NOT_DIR;
    }

    // We don't support OpenFlags.rightDescribe, as service request is not going
    // to be of type fidl.InterfaceRequest<Node> most of the time, but we do send
    // OnOpen event incase of bad flags and if OpenFlags.rightDescribe is passed
    // so that underlying services don't need to handle these flags.
    final unsupportedFlags = ~(OpenFlags.rightReadable |
        OpenFlags.rightWritable |
        OpenFlags.posixWritable |
        OpenFlags.posixExecutable |
        OpenFlags.cloneSameRights);
    if (flags & unsupportedFlags != OpenFlags.$none) {
      return ZX.ERR_NOT_SUPPORTED;
    }

    return ZX.OK;
  }

  @override
  void close() {
    _closed = true;
  }
}
